COTR-105 Implemented shop functionality

disusen opened the pull request:
Implemented basic shop functionality which allows the user to spend his coins on various weapons found in the shop.

disusen committed:
COTR-105 Implemented basic shop functionality

Implemented basic shop functionality

disusen committed:
COTR-105 Minor fixes

Added a label to display errors to user in shop and other minor fixes.

Elijuszek reviewed (changes requested):
- Error when switching weapons
- Files which haven't been touched shouldn't be committed (`.tscn` files)
- When work is of large scope and working across multiple classes changes should be staged into multiple commits (for example two commits for: changes in `Shop.cs` and changes in `WeaponsManager.cs`)

disusen committed:
COTR-105 Moved item count check to WeaponsManager.cs

Moved inventory item count check to WeaponsManager.cs class from WeaponSlot.cs.

Fixed the error when switching weapons.

disusen committed:
COTR-105 Removed unnecessary code

Karolinskis commented on Scenes/Actors/Player/WeaponsManager.cs:
```diff
@@ -60,16 +56,26 @@ public void Initialize(Team.Teams setTeam, Weapon weapon)
     /// Add weapon to currently carrying weapon list
     /// </summary>
     /// <param name="index">Index to put in to</param>
-    public void AddWeapon(Weapon weapon, int index)
+    public bool AddWeapon(Weapon weapon, int index)
```
Due to changing return from `void` to `bool` the summary above this method is incorrect add `<returns>`

Karolinskis commented on Scenes/Actors/Player/WeaponsManager.cs:
```diff
@@ -84,18 +90,13 @@ public Weapon GetCurrentWeapon()
     /// <summary>
     /// Change the current weapon the user is holding
     /// </summary>
-    public void ChangeWeapon()
+    public void ChangeWeapon(int indexToChangeTo)
```
Incorrect summary add <param name="indexToChangeTo">

Karolinskis commented on Scenes/UI/HUD/GUI.cs:
```diff
@@ -194,5 +196,36 @@ private void ChangeItem(int index, Weapon weapon = null)
+        }
+    }
+
+    /// <summary>
+    /// Switches weapon to the first one in weaponsManager.
+    /// </summary>
+    private void ItemPressed1()
+    {
+        SwitchItem(0);
+    }
+
+    /// <summary>
+    /// Switches weapon to the second one in weaponsManager.
+    /// </summary>
+    private void ItemPressed2()
+    {
+        SwitchItem(1);
+    }
+
+    /// <summary>
+    /// Switches weapon to the third one in weaponsManager.
+    /// </summary>
+    private void ItemPressed3()
+    {
+        SwitchItem(2);
+    }
```
This code is unmaintainable. This should probably be one method called `ItemPressed` that you pass and argument in which you tell which button was pressed.

disusen committed:
COTR-105 Fixed code maintainability issues

disusen committed:
COTR-105 Minor fixes

Karolinskis reviewed (changes requested):
If nothing was bought from the store, the player should spawn with no weapon. And the inventory should be empty.

disusen committed:
COTR-105 Fixed bug where switched weapons hitbox fails to appear

disusen committed:
Merge branch 'main' into COTR-105-Add-purchasing-ability

Signed-off-by: disusen <55763781+disusen@users.noreply.github.com>

Karolinskis reviewed (changes requested):
Bug mentioned in https://github.com/Nekompiliuoja/Champions-of-the-Realm/pull/91/commits/f417eedfae778c73d4378ad736635058b80565ac still persists

disusen committed:
COTR-105 Changed all sword scene AnimationPlayer nodes to idle

disusen committed:
COTR-105 Set weapon idle in Melee.cs

disusen committed:
COTR-105 Readability fixes in shop.cs

disusen committed:
COTR-105 typo fix in shop.cs

Karolinskis reviewed (changes requested):
When starting a new game in the debuger an error pops up

Karolinskis commented:
> When starting a new game in the debuger an error pops up

---
E 0:00:08:0416   Godot.NativeInterop.NativeFuncs.generated.cs:1535 @ void Godot.NativeInterop.NativeFuncs.godotsharp_pusherror(Godot.NativeInterop.godot_string& ): Calling Idle method from weapon class
  <C++ Source>   modules/mono/glue/runtime_interop.cpp:1303 @ godotsharp_pusherror()
  <Stack Trace>  Godot.NativeInterop.NativeFuncs.generated.cs:1535 @ void Godot.NativeInterop.NativeFuncs.godotsharp_pusherror(Godot.NativeInterop.godot_string& )
                 GD.cs:351 @ void Godot.GD.PushError(System.String )
                 Weapon.cs:42 @ void ChampionsOfTheRealm.Weapon.Idle()
                 Melee.cs:22 @ void ChampionsOfTheRealm.Melee._Ready()
                 Node.cs:1801 @ Boolean Godot.Node.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CanvasItem.cs:1371 @ Boolean Godot.CanvasItem.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 Node2D.cs:516 @ Boolean Godot.Node2D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 ChampionsOfTheRealm.Weapon_ScriptMethods.generated.cs:61 @ Boolean ChampionsOfTheRealm.Weapon.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 ChampionsOfTheRealm.Melee_ScriptMethods.generated.cs:75 @ Boolean ChampionsOfTheRealm.Melee.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_variant_call_error* , Godot.NativeInterop.godot_variant* )
                 Godot.NativeInterop.NativeFuncs.generated.cs:332 @ void Godot.NativeInterop.NativeFuncs.godotsharp_method_bind_ptrcall(IntPtr , IntPtr , System.Void** , System.Void* )
                 NativeCalls.cs:5859 @ void Godot.NativeCalls.godot_icall_3_655(IntPtr , IntPtr , IntPtr , Godot.NativeInterop.godot_bool , Int32 )
                 Node.cs:600 @ void Godot.Node.AddChild(Godot.Node , Boolean , Godot.Node+InternalMode )
                 Map.cs:127 @ void ChampionsOfTheRealm.Map.SpawnPlayer()
                 Map.cs:53 @ void ChampionsOfTheRealm.Map._Ready()
                 Node.cs:1801 @ Boolean Godot.Node.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CanvasItem.cs:1371 @ Boolean Godot.CanvasItem.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 Node2D.cs:516 @ Boolean Godot.Node2D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 ChampionsOfTheRealm.Map_ScriptMethods.generated.cs:96 @ Boolean ChampionsOfTheRealm.Map.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_variant_call_error* , Godot.NativeInterop.godot_variant* )
                 Godot.NativeInterop.NativeFuncs.generated.cs:332 @ void Godot.NativeInterop.NativeFuncs.godotsharp_method_bind_ptrcall(IntPtr , IntPtr , System.Void** , System.Void* )
                 NativeCalls.cs:5859 @ void Godot.NativeCalls.godot_icall_3_655(IntPtr , IntPtr , IntPtr , Godot.NativeInterop.godot_bool , Int32 )
                 Node.cs:600 @ void Godot.Node.AddChild(Godot.Node , Boolean , Godot.Node+InternalMode )
                 LoadingScreen.cs:96 @ void ChampionsOfTheRealm.LoadingScreen.AddScene()
                 LoadingScreen.cs:71 @ void ChampionsOfTheRealm.LoadingScreen.InitializeResourceLoader()
                 ChampionsOfTheRealm.LoadingScreen_ScriptMethods.generated.cs:54 @ Boolean ChampionsOfTheRealm.LoadingScreen.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_variant_call_error* , Godot.NativeInterop.godot_variant* )

disusen committed:
COTR-105 Fixed error caused by direct Idle() call from weapon.cs

Karolinskis commented on Scenes/UI/HUD/GUI.cs:
```diff
@@ -52,6 +53,13 @@ public override void _Ready()
+        for (int i = 0; i < itemArray.Length; i++)
+        {
+            int currentIndex = i;
+            itemArray[i].Pressed += () => SwitchItem(currentIndex);
```
Refactor this

disusen committed:
COTR-105 Changed for loop to foreach for assigning method to buttons

Karolinskis commented on Scenes/UI/HUD/GUI.cs:
```diff
@@ -55,10 +55,10 @@ public override void _Ready()
-            int currentIndex = i;
-            itemArray[i].Pressed += () => SwitchItem(currentIndex);
+            int ButtonIndex = Array.IndexOf(itemArray, button);
+            button.Pressed += () => SwitchItem(ButtonIndex);
```
This anonymous lambda expression effects code readability in a bad way, please Refactor this part.

Karolinskis commented on Scenes/UI/HUD/GUI.cs:
```diff
@@ -52,6 +53,13 @@ public override void _Ready()
         itemArray[1] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems2");
         itemArray[2] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems3");
+
+        // Assigns each button to SwitchItem method.
+        foreach (Button button in itemArray)
+        {
+            int ButtonIndex = Array.IndexOf(itemArray, button);
+            button.Pressed += () => SwitchItem(ButtonIndex);
+        }
```
Refactor this whole part, don't assign methods through code you can do this through the editor.
![image](https://github.com/Nekompiliuoja/Champions-of-the-Realm/assets/68164180/4d321606-489e-4fe7-a96d-735aebc2e8a9)

disusen committed:
COTR-105 Added signals for inventory buttons to SwitchItem()

Karolinskis commented on Scenes/UI/HUD/GUI.cs:
```diff
@@ -5,6 +5,7 @@ namespace ChampionsOfTheRealm;
 /// </summary>
 public partial class GUI : Control
 {
+    [Signal] public delegate void WeaponSwitchedEventHandler(int index);
```
Why did you add this delegate? It's not needed

disusen committed:
COTR-105 Removed unnecessary code

Karolinskis commented on Scenes/Actors/Player/WeaponsManager.cs:
```diff
@@ -59,17 +55,28 @@ public void Initialize(Team.Teams setTeam, Weapon weapon)
+        }
+        if (index >= weapons.Length)
+        {
+            Array.Resize(ref weapons, index + 1);
```
What's the point of resizing the array when the array is constantly size of 3?

disusen replied:
The array has a maximum of size 3 for items attached through the editor, if any items are added dynamically, like in the shop then array will have to be resized.

Elijuszek commented on Scenes/UI/HUD/GUI.cs:
```diff
@@ -186,13 +188,21 @@ private void ChangeItem(int index, Weapon weapon = null)
         }
-
+        itemArray[index].Disabled = false;
+        itemArray[index].FocusMode = Godot.Control.FocusModeEnum.All;
```
Remove this unnecessary line. By default buttons `focusMode = All`.

Elijuszek replied:
Okay makes sense

Elijuszek reviewed (approved):
LFTM!
